HPCC-28443 Resolve roxie hostnames on demand in ESP services

wangkx opened the pull request:
1. Add new CConnectionSettings to store roxie communication settings.
2. Add new roxie control interfaces to resolve roxie hostnames on demand.
3. Revise ESP services to use the communication settings and new control interfaces.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented on esp/clients/roxiecontrol.cpp:
```diff
@@ -18,12 +18,81 @@
+    *dash = 0;
+    int numberAfterDash = atoi(dash+1);
+    char *lastDot = strrchr(targets, '.');
+    *lastDot = 0;
```
See question above re. why parsing , but if it is necessary, then not good practice to alter memory of owned string like this.

If it's really needed - I assume this range format is used elsewhere. If so, this parsing code should be commoned up in a single hepler caller by both.

jakesmith commented on esp/clients/roxiecontrol.cpp:
```diff
@@ -18,12 +18,81 @@
+const char *CConnectionSettings::nextFQDN()
+{
+    const char *FQDN = FQDNs.item(nextFQDNIndex);
+    ++nextFQDNIndex %= FQDNs.ordinality();
```
related to above - I don't see how this fits in with the schema of a roxie service definition.

jakesmith reviewed (commented):
@wangkx - needs more discussion I think.

The new class (CConnectionSettings) and the maps (e.g. connSettingMap), and conditional code looks like it is in some ways replicating the pre-existing functionality that the smart socket factory implements, but, for example, there should be no ip ranges (or ips) in a roxie service @name that it's parsing.

If it was a new implementation that supported both bare-metal roxie definitions and k8s service definitions and coped with both, thaat would make sense - but the new class is only used by containerized and there's quite of #ifdef _CONTAERINIZED blocks #else bare-metal - which makes the code more complicated and harder to maintain.

Could both bare-metal and containerized use CConnectionSettings ? 

But also, the primary objective here is to avoid cached staatel ep's correct?

It looks like CSmartSocketFactory is already designed to cope with holding hosts (if that's what it is passed via config) and re-resolving them automatically based on a configured timeout (via dnsInterval param)
Was there a reason using that wasn't pursued for resolving this?

wangkx commented on esp/clients/roxiecontrol.cpp:
```diff
@@ -18,12 +18,81 @@
+    if (port == 0)
+        throw makeStringException(-1, "CConnectionSettings: invalid port for service configuration");
+
+    tlsService  = service.getPropBool("@tls");
```
Will fix

wangkx commented on esp/clients/roxiecontrol.cpp:
```diff
@@ -18,12 +18,81 @@
 #define EMPTY_RESULT_FAILURE 1200
 #define SECURE_CONNECTION_FAILURE 1201
 
+CConnectionSettings::CConnectionSettings(IPropertyTree &service)
```
Will add the const.

wangkx commented on esp/smc/SMCLib/TpWrapper.hpp:
```diff
@@ -216,7 +214,11 @@ extern TPWRAPPER_API unsigned getWUClusterInfo(CConstWUClusterInfoArray& cluster
 
 extern TPWRAPPER_API IStringIterator *getContainerTargetClusters(const char* processType, const char* processName);
+#ifndef _CONTAINERIZED
 extern TPWRAPPER_API void initContainerRoxieTargets(MapStringToMyClass<ISmartSocketFactory>& connMap);
```
This version is not called now. The code should be removed when separating the code TpContainer.cpp.

wangkx commented on esp/clients/roxiecontrol.cpp:
```diff
@@ -18,12 +18,81 @@
+void CConnectionSettings::readFQDNs()
+{
+    char *targets = (char *) name.get();
+    char *dash = strrchr(targets, '-');
```
The @name is copied from old code which parses an IP range in CSmartSocketFactory. Do we allow an IP range configured for containerized roxie service?

wangkx commented:
I do agree we need more discussion. 
To my understand, the primary objective here is to avoid cached ep's. I haven't found where the CSmartSocketFactory stores original host settings. So, not sure how to reset the ep's (old IP address) using original host settings in CSmartSocketFactory.
For this PR, the new class (CConnectionSettings) may only be used for containerized roxie. I am still debating whether it should be used to bare metal roxie or not (may be later?). To be used in bare metal roxie, the class has to be upgrade to support many bare metal features (ex. support '|' separated string for multiple host/port settings, etc.).

wangkx commented:
@jakesmith looks like that the name is stored inside the SmartSocketEndpoint. If the dnsInterval is set to 0, no cached IP will be used by the CSmartSocketFactory. So, the cached IP issue may be fixed by changing the dnsInterval to 0 for K8s environment. Any comment?

wangkx commented:
I am testing the 0 dnsInterval. It does not help for the calls from the WsEcl and WsSMC. Will do more research.

wangkx committed:
HPCC-28443 Resolve roxie hostnames on demand in ESP services

1. Set the dnsinterval to 0 for K8s ESP services.
2. Change CSmartSocketFactory::nextEndpoint() to respect the dnsinterval.

Revise based on review:
Move the checkHost call inside nextSmartEndpoint, but make it conditional.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
The ESP services use the roxie connections through the nextEndpoint() call. The nextEndpoint() does not use the dnsInterval. In my latest commit, a new nextEndpointNoCache() is created and used. @afishbeck @jakesmith please check.

afishbeck commented on system/jlib/jsmartsock.cpp:
```diff
@@ -329,6 +329,17 @@ SocketEndpoint& CSmartSocketFactory::nextEndpoint()
     return (ss->ep);
 }
 
+SocketEndpoint& CSmartSocketFactory::nextEndpointNoCache()
```
Perhaps you should just make nextEndpoint() respect the dnsinterval by adding?:

```
ss->checkHost(dnsInterval);
return (ss->ep);
```

afishbeck replied:
And also go back to defaulting the dnsinterval on containerized to 0.

wangkx commented:
@afishbeck I have made the changes related to the nextEndpoint().

jakesmith commented on system/jlib/jsmartsock.cpp:
```diff
@@ -326,6 +326,9 @@ SocketEndpoint& CSmartSocketFactory::nextEndpoint()
         throw createSmartSocketException(0, "smartsocket failed to get nextEndpoint");
+
+    synchronized block(lock);
+    ss->checkHost(dnsInterval);
```
I think should move this inside nextSmartEndpoint, but make it conditional,
e.g. nextSmartEndpoint(bool validate)

call it with true here, and look at other places that call it, which manually call checkHost themselves, so wouldn't need to if nextSmartEndpoint has done already.

wangkx replied:
Moved

jakesmith reviewed (commented):
@wangkx - looks a lot simpler/smaller change.

afishbeck commented on esp/services/ws_ecl/ws_ecl_service.cpp:
```diff
@@ -227,7 +227,7 @@ void initContainerRoxieTargets(MapStringToMyClass<ISmartSocketFactory> &connMap)
             continue;
 
-        Owned<ISmartSocketFactory> sf = new WsEclSocketFactory(service, false, true, nullptr, (unsigned) -1);
+        Owned<ISmartSocketFactory> sf = new WsEclSocketFactory(service, false, true, nullptr, 0);
```
The underlying OS should be doing some level of caching.  Not specific to these particular services, but especially if we do cache though we need to make sure we have retries in which the retry does try to re-resolve the endpoint before new connection attempt. 

I'm not really sure what all the reasons are that IPs will change, but we probably don't want windows where we will fail when they do.

jakesmith replied:
>I'm not really sure what all the reasons are that IPs will change,

I think the main one this issue was related to, was in K8s, where a load-balanced service name resolved to a new IP.

jakesmith replied:
>but we probably don't want windows where we will fail when they do.

I'm not really sure how the smart socket layer is suppsoed to work,
but it looks like it runs a background thread that together with retry/retryInterval, periodically checks the sockets via checkHost(dnsInterval) and actually tried to connect to it (and marks a status).
I'm not sure what checks that status though.

afishbeck replied:
Part of the problem with smart socket fully managing addresses is that it is used in two different ways.. 1. smart socket is directly asked to connect.  2. smart socket is asked to provide an address that another library or function will connect to.

In these cases http routines are used to connect after building a url around the address.

wangkx commented:
@jakesmith could you please review my changes?

jakesmith reviewed (approved):
@wangkx - looks good afaics.

wangkx commented:
@ghalliday this PR is ready to be merged.
